wordpress-mobile/WordPress-Android
# [Tooling] Limit Jetpack supported locales to only Mag16

Internal ref: pxLjZ-7b9-p2

## Why?

While the WordPress app is community-driven and [translated by the community in many locales in GlotPress](https://translate.wordpress.org/projects/apps/android/dev/), the Jetpack app, while based on the same codebase, is an Automattic product whose strings are translated internally by our paid translation vendor, and only in the Mag16 (= the 16 locales we found to be most used)..

We can't really ask the community to translate features that are Jetpack-only like they do for the WordPress app features. But since the Jetpack app shares a lot of screens/features/code with the WordPress app, if we allowed Jetpack to use the translations for all the same locales as WordPress, we could end up with an half-translated Jetpack UI for locales that are not part of the Mag16 (e.g. for someone who has their phone set to `vi`, screens common with WordPress would be translated and in Vietnamese, while Jetpack-only screens would appear in English).

This is why we decided to enforce the Jetpack app binary to only include resource configurations/alternatives for the Mag16 locales — instead of the whole set of locales provided by WordPress' codebase in `src/main/res/values-*`.

Additionally, this helps reduce the app size, by removing unsupported locales in the Jetpack app from the final binary.

## How

This uses `resourceConfigurations` (previously [resConfig](https://developer.android.com/reference/tools/gradle-api/7.3/com/android/build/api/dsl/BaseFlavor#resConfigs(kotlin.collections.Collection))) to [remove non-Mag16 resource alternatives](https://developer.android.com/studio/build/shrink-code.html#unused-alt-resources) for the Jetpack flavor.

This approach also allowed us to [delete the fallback folders we previously created](https://github.com/wordpress-mobile/WordPress-Android/pull/16902) (by having `src/jetpack/res/values-{locale}/` for non-Mag16 be a copy of `src/jetpack/res/values`), and remove the workaround code (creating those fallbacks) from `fastlane/lanes/localization.rb`.

## To Test

#### Test WordPress still includes all resources

 - Run `bundle exec fastlane build_bundle app:wordpress version:res-mag16 flavor:Vanilla buildType:Release skip_lint:true` to build the WordPress aab
 - Run `bundletool dump resources --bundle=build/wpandroid-res-mag16.aab --resource=$res` for the various values of `$res` listed in the ➡️ table below, and ensure you get the expected configurations in return <sup>(†)</sup>

#### Test Jetpack only includes resources for Mag16 locales (but still includes `night`, `landscape`, etc)

 - Run `bundle exec fastlane build_bundle app:jetpack version:res-mag16 flavor:Vanilla buildType:Release skip_lint:true` to build the WordPress aab
 - Run `bundletool dump resources --bundle=build/jpandroid-res-mag16.aab --resource=$res` for the various values of `$res` listed in the ➡️ table below, and ensure you get the expected configurations in return <sup>(†)</sup>

<details><summary><sup>(†)</sup> You can use the following shell script to help you with these tests</summary>

```sh
#!/bin/sh

aab=${1:-build/jpandroid-res-mag16.aab}

# Resource => What to expect
# string/app_title => (default), mag16
# dimen/prologue_promo_title_size => (default), smallest_screen_width_dp: 600, orientation: LAND
# string/reader_author_with_blog_name => (default), layout_direction: RTL
# color/placeholder => (default), ui_night_mode: NIGHT

for res in string/app_title dimen/prologue_promo_title_size string/reader_author_with_blog_name color/placeholder; do
  bundletool dump resources --bundle=$aab --resource=$res
done
```
</details>

<details><summary>➡️ Resources to test for <tt>bundletool dump resources</tt> for each app</summary>

And the corresponding expected results (in addition to `(default)`) for each app flavor:

| `$res` | WordPress | Jetpack |
|---|---|---|
| `string/app_title` | 53 locales, including non-Mag16 ones like `vi` | Only Mag16 |
| `dimen/prologue_promo_title_size` | `smallest_screen_width_dp: 600`, `orientation: ORIENTATION_LAND` | _same as WP_
| `string/reader_author_with_blog_name` | `layout_direction: LAYOUT_DIRECTION_RTL` | _same as WP_ |
| `color/placeholder` | `ui_night_mode: UI_MODE_NIGHT_NIGHT` | _same as WP_ |
</details>

#### Test that the new constency check in fastlane works as expected

 - Run `./gradlew printResourceConfigurations`
 - Validate that the last lines prints an array of the Mag16 locales for `jetpack`, and an empty array for `wordpress`
 - Cut a test branch from this PR's branch
 - Run `bundle exec fastlane download_translations`
 - Check that in the logs printed while running that lane, `printResourceConfigurations` is called twice:
    - Once for WordPress (before downloading the WP translations), ending with a message ``No `resourceConfigurations` field set on the `wordpress` flavor to limit the locales. Nothing to check.``
    - Once for Jetpack (before downloading the JP translations), ending with a message like ``The `resourceConfigurations` field set in `build.gradle` for the `jetpack` flavor matches what is set in our Fastfile. All is good!``
 - Modify either the `resourceConfigurations` from `build.gradle`, or the `JP_APP_LOCALES` array to add/remove locales, run the test again, and ensure that this time the lane stops with an error to complain about the configuration inconsistency for Jetpack.
 - Delete your test branch, **both** from your local git **and the remote** — since that test ends up doing a `git push`

#### Test on device

 - Install the Jetpack `aab` on your test device
 - Set your test device to a non-Mag16 locale like Vietnamese
 - Ensure the Jetpack app still shows its UI in English, not Vietnamese
 - Go to Me screen > App Settings > Interface Language and check that the list only lists Mag16 (+ English) and does not contain e.g. Vietnamese.
